### ISSUE TITLE
Demo project: run command lines via custom Python server 

### ISSUE THREAD

**Issue by daabr** [autokitteh/kittehub#131]
AutoKitteh doesn't have a mechanism (i.e. integration) to execute arbitrary command lines directly in the server's OS because it's too easy to abuse it, but users can still achieve this through a separate server of their own, which communicates with AutoKitteh and supports any custom implementation details they desire.

Context: https://github.com/autokitteh/autokitteh/issues/861#issuecomment-2511488838

High level design:

1. Implement a custom Python web server (e.g. with [Flask](https://flask.palletsprojects.com/en/stable/) or [FastAPI](https://fastapi.tiangolo.com/)?) that runs as an independent process on a Windows machine in the user's private network, accepts HTTP requests to run commands-lines, uses the [subprocess](https://docs.python.org/3/library/subprocess.html) Python module, and returns the results

2. AutoKitteh server: self-hosted in the same private network (more setup work but less security risks), or https://app.autokitteh.cloud/ (no AutoKitteh setup, but requires HTTP tunneling for the custom Python web server, and thinking about security)

3. AutoKitteh project:
   - Any desirable project trigger(s)
   - Project var: the custom Python server's base URL
   - Workflow Python code: parse the triggering event's details (TBD), or simply use hard-coded parameters, to construct and send an HTTP request to the custom Python server

[See this example of using a project var, and HTTP in AutoKitteh](https://github.com/autokitteh/kittehub/blob/5ae61242bd93c06293973182f7f4584d3f577063/samples/http/autokitteh.yaml#L9)

This demo enables a bunch of interesting and versatile use-cases, e.g. controlling k8s by running `kubectl` (even though we plan to provide a built-in integration for that), running UI automation programs, controlling home automation agents, etc. etc.

**Comment by dolby360** [comment]
NP assign me.

**Comment by dolby360** [comment]
Thanks I'm waiting for this issue to be solved so I can start
https://github.com/autokitteh/autokitteh/issues/915

**Comment by daabr** [comment]
The fix for https://github.com/autokitteh/autokitteh/issues/915 is now merged. I'm also going to do additional testing do ensure that the user experience in Docker is acceptable.

**Comment by dolby360** [comment]
I've been investigating this for a few days, and it seems there isn't a straightforward solution (or one that doesn't involve many workarounds) to enable communication from WSL to Windows, which is a simpler case. They don't share the same localhost. I did manage to establish communication from Docker to WSL, but I feel it misses the point.
What do you think?
@daabr

**Comment by daabr** [comment]
@dolby360 making sure I understand correctly:

1. Communication between AutoKitteh in Docker (on Windows?) and WSL (on the same Windows machine?) does work
   - If so, do you have any interesting notes about enabling/testing this?
2. Unrelated to AutoKitteh(?), communication between WSL and the underlying Windows OS - no straightforward solution / involves many workarounds

Did I get this right?

If so, my clarification is that my idea was to run the Python web server as a native Windows process, regardless of whether AK is running on WSL/Docker on the same machine, or not. For that matter, you could also write a lightweight .NET web server instead of using Python.

I **assume** that it's possible to expose ports between Docker and/or WSL and Windows. Or is that not the case?

**Comment by dolby360** [comment]
@daabr 
WSL and Docker coexist in the same environment, so they’re essentially the same for me in this context.
Communication between WSL and Windows isn’t straightforward—I had to use a workaround to make it function properly.
Let me explain. You can’t simply send requests to localhost from WSL, listening on Windows at `0.0.0.0:port`, and expect it to work. The reverse is possible, but when sending requests from Windows to WSL, you won’t use localhost. Instead, you’ll send them to an IP address, which you can find by opening PowerShell and running `wsl hostname -I`.

Regarding your comment:

_My idea was to run the Python web server as a native Windows process, regardless of whether AK is running on WSL/Docker on the same machine, or not._

I’m not sure I understand what we’re building. Are we just creating a server to receive and process requests? If so, how is this different from the many examples available online?

**Comment by daabr** [comment]
Thanks for the clarification @dolby360! As for the last question: the idea was to use this web server as a remote agent of AutoKitteh - so the overall operation, orchestration, and integration with other services is still managed in AutoKitteh, but it also has a "backdoor" to interact with a Windows machine, unrelated to where AK and that Windows machine are.

But I get your point, and I'll try to dig deeper into running AK on Windows.

Incidentally, we started working on integrations with Microsoft 365, which requires an Azure account, so it's in the realm of possibility to run on Windows VMs now in some capacity.

**Comment by dolby360** [comment]
@daabr I believe we should put this task on hold for now. Presenting a demo with awkward workarounds might do more harm than good.

**Comment by daabr** [comment]
Makes sense, but of course we'll continue to pursue this from the other directions, and let you know when we have notable updates.
